subprojects/core/src/test/groovy/org/gradle/cache/internal/DefaultCacheAccessTest.groovy

```diff
@@ -631,4 +634,119 @@ class DefaultCacheAccessTest extends ConcurrentSpec {
+
+        cleanup:
+        access?.close()
+    }
+
+    def "throws InvalidCacheReuseException when cache value type differs"() {
+        def access = newAccess(None)
+
+        when:
+        access.newCache(new PersistentIndexedCacheParameters('cache', String.class, Integer.class))
+        access.newCache(new PersistentIndexedCacheParameters('cache', String.class, String.class))
+
+        then:
+        thrown(DefaultCacheAccess.InvalidCacheReuseException)
+
+        cleanup:
+        access?.close()
+    }
+
+    def "throws InvalidCacheReuseException when cache key type differs"() {
+        def access = newAccess(None)
+
+        when:
+        access.newCache(new PersistentIndexedCacheParameters('cache', String.class, Integer.class))
+        access.newCache(new PersistentIndexedCacheParameters('cache', Integer.class, Integer.class))
+
+        then:
+        thrown(DefaultCacheAccess.InvalidCacheReuseException)
+
+        cleanup:
+        access?.close()
+    }
+
+    def "throws InvalidCacheReuseException when cache decorator differs"() {
+        def access = newAccess(None)
+        def decorator = Mock(CacheDecorator)
+        lockManager.lock(lockFile, mode(Exclusive), "<display-name>") >> lock
+        decorator.decorate(_, _, _, _, _) >> { String cacheId, String cacheName, MultiProcessSafePersistentIndexedCache persistentCache, CrossProcessCacheAccess crossProcessCacheAccess, AsyncCacheAccess asyncCacheAccess ->
+            persistentCache
+        }
+
+        when:
+        access.newCache(new PersistentIndexedCacheParameters('cache', String.class, Integer.class))
+        access.newCache(new PersistentIndexedCacheParameters('cache', String.class, Integer.class).cacheDecorator(decorator))
+
+        then:
+        thrown(DefaultCacheAccess.InvalidCacheReuseException)
+
+        cleanup:
+        access?.close()
+    }
+
+    def "returns the same cache object when cache decorator match"() {
+        def access = newAccess(None)
+        def decorator1 = Mock(CacheDecorator)
+        def decorator2 = Mock(CacheDecorator)
+        lockManager.lock(lockFile, mode(Exclusive), "<display-name>") >> lock
+        decorator1.decorate(_, _, _, _, _) >> { String cacheId, String cacheName, MultiProcessSafePersistentIndexedCache persistentCache, CrossProcessCacheAccess crossProcessCacheAccess, AsyncCacheAccess asyncCacheAccess ->
+            persistentCache
+        }
+        decorator2.decorate(_, _, _, _, _) >> { String cacheId, String cacheName, MultiProcessSafePersistentIndexedCache persistentCache, CrossProcessCacheAccess crossProcessCacheAccess, AsyncCacheAccess asyncCacheAccess ->
+            persistentCache
+        }
+
+        when:
+        def cache1 = access.newCache(new PersistentIndexedCacheParameters('cache', String.class, Integer.class).cacheDecorator(decorator1))
+        def cache2 = access.newCache(new PersistentIndexedCacheParameters('cache', String.class, Integer.class).cacheDecorator(decorator2))
```

Comment:
I think the simplest thing to do is to just implement `CacheDecorator` here and use the same instance (vs using a Mock).  That matches what we're doing in the real build.

I think checking for the same instance is fine.